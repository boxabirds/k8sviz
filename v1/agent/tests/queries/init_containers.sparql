# Query: Deployment with init containers for setup tasks
# This query retrieves information about init containers and their usage

PREFIX k8s: <http://kubernetes.io/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>

SELECT ?resource ?property ?value ?comment
WHERE {
  # Get init container information
  {
    k8s:hasInitContainer rdfs:label ?label ;
                        rdfs:comment ?comment .
  }
  
  # Get container properties relevant to init containers
  UNION
  {
    ?property rdfs:domain k8s:Container ;
              rdfs:label ?propertyLabel ;
              rdfs:comment ?propertyComment .
    FILTER(CONTAINS(STR(?property), "container"))
  }
  
  # Get Pod lifecycle information
  UNION
  {
    k8s:Pod rdfs:comment ?podComment .
    FILTER(CONTAINS(LCASE(STR(?podComment)), "container"))
  }
  
  # Get best practices for init containers
  UNION
  {
    ?practice a k8s:BestPractice ;
              rdfs:comment ?practiceComment .
    FILTER(CONTAINS(LCASE(STR(?practiceComment)), "init") || 
           CONTAINS(LCASE(STR(?practiceComment)), "startup"))
  }
}