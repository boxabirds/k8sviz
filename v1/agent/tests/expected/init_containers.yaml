apiVersion: apps/v1
kind: Deployment
metadata:
  name: app-with-init
spec:
  replicas: 3
  selector:
    matchLabels:
      app: myapp
  template:
    metadata:
      labels:
        app: myapp
    spec:
      initContainers:
      # Wait for database to be ready
      - name: wait-for-db
        image: busybox:1.35
        command: ['sh', '-c', 'until nc -z db-service 5432; do echo waiting for db; sleep 2; done']
        resources:
          requests:
            cpu: "50m"
            memory: "64Mi"
          limits:
            cpu: "100m"
            memory: "128Mi"
      # Run database migrations
      - name: db-migration
        image: myapp/migration:latest
        env:
        - name: DATABASE_URL
          valueFrom:
            secretKeyRef:
              name: db-secret
              key: url
        resources:
          requests:
            cpu: "100m"
            memory: "256Mi"
          limits:
            cpu: "500m"
            memory: "512Mi"
      # Download and prepare static assets
      - name: prepare-assets
        image: myapp/assets:latest
        command: ['sh', '-c', 'cp -r /assets/* /app-assets/']
        volumeMounts:
        - name: assets
          mountPath: /app-assets
        resources:
          requests:
            cpu: "100m"
            memory: "128Mi"
          limits:
            cpu: "200m"
            memory: "256Mi"
      containers:
      - name: app
        image: myapp/app:v1.0.0
        ports:
        - containerPort: 8080
        env:
        - name: DATABASE_URL
          valueFrom:
            secretKeyRef:
              name: db-secret
              key: url
        volumeMounts:
        - name: assets
          mountPath: /app/static
          readOnly: true
        resources:
          requests:
            cpu: "250m"
            memory: "512Mi"
          limits:
            cpu: "1000m"
            memory: "1Gi"
        livenessProbe:
          httpGet:
            path: /health
            port: 8080
          initialDelaySeconds: 30
          periodSeconds: 10
        readinessProbe:
          httpGet:
            path: /ready
            port: 8080
          initialDelaySeconds: 10
          periodSeconds: 5
        startupProbe:
          httpGet:
            path: /startup
            port: 8080
          failureThreshold: 30
          periodSeconds: 10
      volumes:
      - name: assets
        emptyDir: {}